src/Util/Asserts.h

```diff
@@ -13,24 +13,40 @@
 #define CONVERT_TO_STRING(X) #X
 #define CONVERT_EXPR_TO_STRING(X) CONVERT_TO_STRING(X)
 
-#define OWSAssert(X) \
-if (!(X)) { \
-NSLog(@"Assertion failed: %s", CONVERT_EXPR_TO_STRING(X)); \
-NSAssert(0, @"Assertion failed: %s", CONVERT_EXPR_TO_STRING(X)); \
-}
-
-// OWSAssert() should be used in Obj-C and Swift methods.
-// OWSCAssert() should be used in free functions.
-#define OWSCAssert(X) \
-if (!(X)) { \
-NSLog(@"Assertion failed: %s", CONVERT_EXPR_TO_STRING(X)); \
-NSCAssert(0, @"Assertion failed: %s", CONVERT_EXPR_TO_STRING(X)); \
-}
+// OWSAssert() and OWSFail() should be used in Obj-C methods.
+// OWSCAssert() and OWSCFail() should be used in free functions.
+
+#define OWSAssert(X)                                                                                                   \
+    if (!(X)) {                                                                                                        \
+        NSLog(@"%s Assertion failed: %s", __PRETTY_FUNCTION__, CONVERT_EXPR_TO_STRING(X));                             \
+        NSAssert(0, @"Assertion failed: %s", CONVERT_EXPR_TO_STRING(X));                                               \
+    }
+
+#define OWSCAssert(X)                                                                                                  \
+    if (!(X)) {                                                                                                        \
+        NSLog(@"%s Assertion failed: %s", __PRETTY_FUNCTION__, CONVERT_EXPR_TO_STRING(X));                             \
+        NSCAssert(0, @"Assertion failed: %s", CONVERT_EXPR_TO_STRING(X));                                              \
+    }
+
+#define OWSFail(X)                                                                                                     \
```

Comment:
> On a more radical note, what if we defined these in non-debug, and changed the NSLog to DDLogError so that we can see them in production logs (and eventually they could be reported).

I'd love to do that logging in production - but not actually use the asserts.  I think it's key for stability that our code fails gracefully in production.  I'm pretty sure we're rolling past asserts like crazy in production right now. 

And the NSLog usage here isn't great. But I haven't converted them to DDLogError() because I was worried about PII leaking into the production logs.  

My long-term hope is to unify:

* Logging
* Asserts
* Analytics
* Return values from these macros so that we can be more explicit about how to handle deviations from these invariants in production. 

I'm hoping that we could land on a _small_ handful of macros like:

```
// This macro would DDLogError(), Assert in DEBUG, send an analytics event, return a value that indicates the assert failed... all in a tidy little consistent bit of code.
if (!DDCriticalAssertWAnalytics(message != nil, @"Incoming message failed to decrypt")) {
   return;
}
```

WDYT?